src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,1389 @@
+            }
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces);
+            comp.VerifyDiagnostics(
+                // (8,26): error CS1996: Cannot await in the body of a lock statement
+                //             using await (var x = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitInLock, "var x = new C()").WithLocation(8, 26)
+                );
+        }
+
+        [Fact]
+        public void TestInCatchBlock()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    public static async System.Threading.Tasks.Task Main()
+    {
+        try
+        {
+            System.Console.Write(""try "");
+            throw new System.ArgumentNullException();
+        }
+        catch (System.ArgumentNullException)
+        {
+            using await (var x = new C())
+            {
+                System.Console.Write(""using "");
+                return;
+            }
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        System.Console.Write(""dispose"");
+        return System.Threading.Tasks.Task.CompletedTask;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "try using dispose");
+        }
+
+        [Fact]
+        public void MissingAwaitInAsyncMethod()
+        {
+            string source = @"
+class C
+{
+    public static async System.Threading.Tasks.Task Main()
+    {
+        System.Action a = () =>
+        {
+            System.Action b = async () => { await local(); }; // this await doesn't count towards Main method
+        };
+
+        async System.Threading.Tasks.Task local()
+        {
+            await local(); // neither does this one
+        }
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source);
+            comp.VerifyDiagnostics(
+                // (4,53): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     public static async System.Threading.Tasks.Task Main()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "Main").WithLocation(4, 53)
+                );
+        }
+
+        [Fact]
+        public void MissingAwaitInAsyncMethod2()
+        {
+            string source = @"
+class C
+{
+    public static void Main()
+    {
+        System.Action lambda1 = async () =>
+        {
+            System.Action b = async () => { await local(); }; // this await doesn't count towards lambda
+        };
+
+        System.Action lambda2 = async () => await local2(); // this await counts towards lambda
```

Comment:
>await local2() [](start = 44, length = 14)

Shouldn't we be testing ```using await``` here?

Reply:
We can't here. The purpose for this lambda is to have an expression body, but async using is a statement.

---
In reply to: [162705223](https://github.com/dotnet/roslyn/pull/23961#discussion_r162705223) [](ancestors = 162705223)